Infosec overview document (#1)
teoparvanov/openfl: infosec-orverview.md

### Purpose of this document

> This document is intended to help InfoSec analysis processes of the collaborators involved in the initial stages of OpenFL federations by summarizing key points of the code functionality.

Reviewer: ```suggestion
This document provides the information needed when evaluating OpenFL for real world deployment in highly sensitive environments. The target audience is InfoSec reviewers who need detailed information about code contents, communication traffic, and potential exploit vectors.
```

#### Network Connectivity Overview
OpenFL federations use a hub-and-spoke topology between _collaborator_ clients that generate model parameter updates from their data and the _aggregator_ server that combines their training updates into new models. Key details about this functionality are:

> * Connections are made using request/response gRPC connections.

Reviewer 2: ```suggestion
* Connections are made using request/response gRPC connections [[ref](https://grpc.io/docs/what-is-grpc/core-concepts/)].
```

> * The _aggregator_ listens for connections on a single port, explicitly defined in the FL plan (f.e. 50051), so all _collaborators_ must be able to send outgoing traffic to this port.

Reviewer 2: ```suggestion
* The _aggregator_ listens for connections on a single port (usually decided by the experiment admin), and is explicitly defined in the FL plan (f.e., `50051`), so all _collaborators_ must be able to send outgoing traffic to this port.
```

> * All connections are initiated by the _collaborator_.

Reviewer 2: ```suggestion
* All connections are initiated by the _collaborator_, i.e., a `pull` architecture [[ref](https://karlchris.github.io/data-engineering/data-ingestion/push-pull/#pull)].
```

* The _collaborator_ does not open any listening sockets.

> * Connections are secured using mutually-authenticated TLS.

Reviewer 2: ```suggestion
* Connections are secured using mutually-authenticated TLS [[ref](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/)].
```

* Each request response pair is done on a new TLS connection.

> * The PKI for OpenFL federations can be created using openssl tools, with the organization hosting the _aggregator_ usually acting as the Certificate Authority (CA) that verifies each identity before signing.

Reviewer: We now use the cryptography package instead of OpenSSL

```suggestion
* The PKI for federations can be created using the [OpenFL CLI](https://openfl.readthedocs.io/en/latest/about/features_index/taskrunner.html#step-2-configure-the-federation). OpenFL internally leverages Python's cryptography module. The organization hosting the _aggregator_ usually acts as the Certificate Authority (CA) and verifies each identity before signing.
```

* Currently, the _collaborator_ polls the _aggregator_ at a fixed interval. We have had a request to enable client-side configuration of this interval and hope to support that feature soon.
* Connection timeouts are set to gRPC defaults.
* If the _aggregator_ is not available, the _collaborator_ will retry connections indefinitely. This is currently useful so that we can take the aggregator down for bugfixes without _collaborator_ processes exiting.

#### Overview of Contents of Network Messages

> Network messages are well defined protobufs which can be found in [aggregator.proto](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/aggregator.proto) and [base.proto](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/base.proto).

Reviewer 2: ```suggestion
Network messages are well defined protobufs which can be found in the following files:
- [aggregator.proto](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/aggregator.proto) 
- [base.proto](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/base.proto)

```

Key points about the network messages/protocol:

> * No private data is ever sent to the _aggregator_

Reviewer 2: ```suggestion
* No private data is ever sent to the _aggregator_.
```

Author: N/A as I've removed this bullet per Brandon's comment above.

* No executable code is ever sent to the _collaborator_. All algorithms come with the OpenFL package and the custom FL workspace.

> * The _collaborator_ typically requests the FL tasks to execute from the aggregator via the `GetTasksRequest` message

Reviewer 2: ```suggestion
* The _collaborator_ typically requests the FL tasks to execute from the aggregator via the `GetTasksRequest` message [[ref](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/aggregator.proto#L34)].
```

> * The _aggregator_ reads the FL plan and returns a `GetTasksResponse` which includes metadata (`Tasks`) about the Python functions to be invoked by the collaborator (the code being installed locally as part of a pre-distributed workspace bundle)

Reviewer 2: ```suggestion
* The _aggregator_ reads the FL plan and returns a `GetTasksResponse` [[ref](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/aggregator.proto#L45)] which includes metadata (`Tasks`) about the Python functions to be invoked by the collaborator (the code being installed locally as part of a pre-distributed workspace bundle).
```

> * The _collaborator_ then uses its TaskRunner framework to execute the FL tasks on the locally available data, producing output tensors such as model weights or metrics

Reviewer 2: ```suggestion
* The _collaborator_ then uses its `TaskRunner` framework [[ref](https://openfl.readthedocs.io/en/latest/about/features_index/taskrunner.html)] to execute the FL tasks on the locally available data, producing output tensors such as model weights or metrics.
```

> * During task execution, the _collaborator_ may additionally request tensors from the aggregator via the `GetAggregatedTensor` RPC method

Reviewer 2: ```suggestion
* During task execution, the _collaborator_ may additionally request tensors from the aggregator via the `GetAggregatedTensor` RPC method [[ref](https://openfl.readthedocs.io/en/latest/reference/_autosummary/openfl.transport.grpc.aggregator_server.AggregatorGRPCServer.html#openfl.transport.grpc.aggregator_server.AggregatorGRPCServer.GetAggregatedTensor)].
```

> * Upon task completion, the _collaborator_ transmits the results by emitting a `SendLocalTaskResults` call which contains `NamedTensor` objects that encode model weight updates or ML metrics such as loss or accuracy (among others).

Reviewer 2: ```suggestion
* Upon task completion, the _collaborator_ transmits the results by emitting a `SendLocalTaskResults` call [[ref](https://openfl.readthedocs.io/en/latest/reference/_autosummary/openfl.transport.grpc.aggregator_server.AggregatorGRPCServer.html#openfl.transport.grpc.aggregator_server.AggregatorGRPCServer.SendLocalTaskResults)] which contains `NamedTensor` objects that encode model weight updates or ML metrics such as loss or accuracy (among others).
```

#### Testing a Collaborator

> In order to test a _collaborator_, the _aggregator_ can be set to not aggregate updates from specific _collaborators_ so that they can test functionality without impacting the running model training/validation. We hope this enables InfoSec runtime analysis against the live _aggregator_ without the need to access any private data on the _collaborator_ machine running the test. To do this, please coordinate with the OpenFL aggregator admins.

Author: As of [the April 15th OpenFL nightly build](https://pypi.org/project/openfl-nightly/1.9.0.dev20250415/), we've added a [`no-op`](https://github.com/securefederatedai/openfl/tree/develop/openfl-workspace/no-op#overview) workspace template, which can be used in conjunction with the recently introduced `fx collaborator ping` command to test collaborator/aggregator connectivity without running any FL tasks or accessing private data.

PS: this functionality will be officially available in the upcoming OpenFL 1.9 release

Reviewer 3: A few thoughts:

Maybe " No private data is ever sent to the _aggregator_" should be removed, as this is not guaranteed by the messaging protocol itself. I could directly embed the data into a model sized tensor and send as a "model update".

I would expand the next bullet to read *something like*, "No executable code is ever sent to the _collaborator_. All code to be executed resides in the combination of the OpenFL package and custom FL workspace. These codes, along with the FL plan file that defines the classes and initial parameters to be used, is available for review before the federation runs in order that all possible execution is understood before it occurs."

> In order to test a _collaborator_, the _aggregator_ can be set to not aggregate updates from specific _collaborators_ so that they can test functionality without impacting the running model training/validation. We hope this enables InfoSec runtime analysis against the live _aggregator_ without the need to access any private data on the _collaborator_ machine running the test. To do this, please coordinate with the OpenFL aggregator admins.

Reviewer 2: ```suggestion
There is a "no-op" mode in OpenFL (available in versions `>=1.9`) which tests the network connection between the _aggregator_ and each _collaborator_ without performing any computational task. More details can be found in [this link](https://github.com/securefederatedai/openfl/tree/develop/openfl-workspace/no-op#overview).
```
